js/ui/payg.js

```diff
@@ -0,0 +1,195 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+//
+// Copyright (C) 2018 Endless Mobile, Inc.
+//
+// Licensed under the GNU General Public License Version 2
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+const Gettext = imports.gettext;
+const Gio = imports.gi.Gio;
+const GLib = imports.gi.GLib;
+const GObject = imports.gi.GObject;
+const Lang = imports.lang;
+const Mainloop = imports.mainloop;
+
+const PaygManager = imports.misc.paygManager;
+
+const Main = imports.ui.main;
+const Tweener = imports.ui.tweener;
+
+var SUCCESS_DELAY_SECONDS = 3;
+
+var SPINNER_ICON_SIZE_PIXELS = 16;
+var SPINNER_ANIMATION_DELAY_SECS = 1.0;
+var SPINNER_ANIMATION_TIME_SECS = 0.3;
+
+var UnlockStatus = {
+    NOT_VERIFYING: 0,
+    VERIFYING: 1,
+    FAILED: 2,
+    TOO_MANY_ATTEMPTS: 3,
+    SUCCEEDED: 4,
+};
+
+var UnlockUi = new Lang.Class({
+    Name: 'PaygUnlockUi',
+    Extends: GObject.Object,
+
+    _init: function() {
+        this.parent();
+        this._clearTooManyAttemptsId = 0;
+        this.connect('destroy', this._onDestroy.bind(this));
+    },
+
+    updateNextButtonSensitivity: function() {
+
+        let sensitive = this.validateCurrentCode() &&
+            this.verificationStatus != UnlockStatus.VERIFYING &&
+            this.verificationStatus != UnlockStatus.TOO_MANY_ATTEMPTS;
+
+        this.applyButton.reactive = sensitive;
```

Comment:
Done. I looked at doing a legit GInterface in JS. It should be possible but it adds some pretty horrific boilerplate code (not quite but surprisingly close to implementing interfaces in GObject/C) since there are 8 properties.

In the end, the best option seems to be adding documentation in case anyone else later subclasses again at the risk of aliasing code with docs.